cypress/support/commands.js

```diff
@@ -126,16 +126,19 @@ Cypress.Commands.add('createApplication', (application, userId) => {
 })
 
 // Helper function to create a new api key programmatically
-Cypress.Commands.add('createApiKey', (mode, name, apiKey) => {
+Cypress.Commands.add('createApiKey', (entity, entityId, apiKey) => {
   const baseUrl = Cypress.config('baseUrl')
   cy.getAccessToken(accessToken => {
     cy.request({
       method: 'POST',
-      url: `${baseUrl}/api/v3/${mode}/${name}/api-keys`,
+      url: `${baseUrl}/api/v3/${entity}/${entityId}/api-keys`,
       body: { ...apiKey },
       headers: {
         Authorization: `Bearer ${accessToken}`,
       },
+    }).then(response => {
+      const apiKeyId = response.body && response.body.id
+      cy.wrap(apiKeyId).as(`${entity}_apiKeyId`)
```

Comment:
I had an issue getting the `API key id` to navigate to the edit page directly.

In my first attempt I have implemented this using a callback function, which worked.

Then I have read about aliases and so I decided to set and use my API key ids like that.
Unfortunately, now I have to call `createApiKey` in `beforeEach` test, because `before` only runs once and Cypress will remove all of the aliases between tests.

I have also read that the response could potentially be saved using [writeFile](https://docs.cypress.io/api/commands/writefile.html#JSON) or [Cypress.env](https://docs.cypress.io/api/cypress-api/env.html#Syntax)

Let me know if my alias solution is not great here and what are my other options. Thanks!

Reply:
While this solves the issue, such implementation also has some issues that might cause problems later. 

1. Implicit alias `${entity}_apiKeyId` than one just have to know. I think we should avoid referencing aliases in between files.
2. It only returns the id. After adding https://github.com/TheThingsNetwork/lorawan-stack/issues/3481 we will add assertions on copying the api key. We should approach this in a more generic way, at least returning the whole resulting object.
3. Alias references are removed between tests. In this case you are forced to re-create the api key before each test. This is not optimal.

I think that adding a callback is a better options. Overall, it might be helpful for other API commands to allow retrieving response objects:

```js
// commands.js

import { noop } from 'lodash'

// Helper function to create a new api key programmatically
Cypress.Commands.add('createApiKey', (entity, entityId, apiKey, cb = noop) => {
  const baseUrl = Cypress.config('baseUrl')
  cy.getAccessToken(accessToken => {
    cy.request({
      method: 'POST',
      url: `${baseUrl}/api/v3/${entity}/${entityId}/api-keys`,
      body: { ...apiKey },
      headers: {
        Authorization: `Bearer ${accessToken}`,
      },
    }).then(resp => {
      cb(resp.body)
    })
  })
})
```

And then using it:

```js
// edit.spec.js

describe('...', () => {
  let apiKeyId = null // not needed if `Cypress.config` is used in `before` hook.

  before(() => {
    cy.createApiKey(entity, applicationId, apiKey, key => {
      apiKeyId = key.id
      // Cypress.config('apiKeyId', key.id) is also possible as config is scoped within the spec file.
    })
  })

  beforeEach(() => {
    cy.wrap(apiKeyId).as('apiKeyId') // not needed if `Cypress.config` is used in `before` hook.
  })

  it('...', () => {
    cy.get('@apiKeyId').then(id => {
      cy.visit(`.../${id}`)
    })
    // or cy.visit(`.../${Cypress.config('apiKeyId')}`)
  })
})

```

Reply:
It seems that my comment with detailed explanation why this is not the best approach got deleted 🤷‍♂️🤷‍♂️🤷‍♂️.

tldr: 
1. Sharing aliases between spec/utility files might cause us problems later. 
2. Why returning only the id? After https://github.com/TheThingsNetwork/lorawan-stack/issues/3481 we will add assertions on copying the api key - in this case we will need the complete response object.
3. This is not scalable. Imagine if I need to access response objects for other API commands.

IMO, passing callback is a better option.

```js
// commands.js

import { noop } from 'lodash'

// Helper function to create a new api key programmatically
Cypress.Commands.add('createApiKey', (entity, entityId, apiKey, cb = noop) => {
  const baseUrl = Cypress.config('baseUrl')
  cy.getAccessToken(accessToken => {
    cy.request({
      method: 'POST',
      url: `${baseUrl}/api/v3/${entity}/${entityId}/api-keys`,
      body: { ...apiKey },
      headers: {
        Authorization: `Bearer ${accessToken}`,
      },
    }).then(response => cb(response.body))
  })
})

// edit.spec.js

let apiKeyId = null // not needed if `Cypress.config` is used in `before` hook.

before(() => {
   cy.createApiKey(entity, applicationId, apiKey, key => {
     apiKeyId = key.id
     // or `Cypress.config('apiKeyId', key.id)`
   })
})

beforeEach(() => {
  cy.wrap(apiKeyId).as('apiKeyId') // not needed if `Cypress.config` is used.
})

it('...', () => {
  cy.get('@apiKeyId').then(apiKeyId => {
	cy.visit(`.../${apiKeyId}`)
  })
 // or cy.visit(`.../${Cypress.config('apiKeyId')}`)
})
```

ping me in slack if you need more details on this

Reply:
Thanks a lot for providing details on how to better approach this! I've decided to go with `Cypress.config`.